### PR TITLE
Implement pretty-printing of combined `enum.Flag` instances

### LINKED ISSUE #3707: 6.82.0 does not build with Python 3.8
Hypothesis 6.82.0 does not build with Python 3.8:

```
Traceback (most recent call last):
  File "setup.py", line 31, in <module>
    SOURCE = str(local_file("src"))
  File "setup.py", line 28, in local_file
    return Path(__file__).parent.joinpath(name).relative_to(Path.cwd())
  File "/usr/pkg/lib/python3.8/pathlib.py", line 908, in relative_to
    raise ValueError("{!r} does not start with {!r}"
ValueError: 'src' does not start with '/tmp/devel/py-hypothesis/work/hypothesis-6.82.0'
```

That's because Python 3.8 sets `__file__` as a relative path.

Here is my fix:

```
--- setup.py.orig	2023-07-29 15:30:00.000000000 +0000
+++ setup.py
@@ -25,7 +25,7 @@ if sys.version_info[:2] < (3, 8):
 
 
 def local_file(name):
-    return Path(__file__).parent.joinpath(name).relative_to(Path.cwd())
+    return Path(__file__).absolute().parent.joinpath(name).relative_to(Path.cwd())
 
 
 SOURCE = str(local_file("src"))
```

### LINKED ISSUE #3709: Generating mixed value for enum.Flag causes pretty-printer exception
A self-contained reproduction:

```
import enum

from hypothesis import example, given
from hypothesis import strategies as st


class MyFlag(enum.Flag):
    ONE = 1
    TWO = 2


@given(st.sampled_from(MyFlag))
@example(MyFlag.ONE | MyFlag.TWO)
def test_pretty_printing(flag: MyFlag):
    assert isinstance(flag, MyFlag)
```

Tests fail with the following error message:

```
test_hypothesis_bug.py::test_pretty_printing Failed: [undefined]TypeError: can only concatenate str (not "NoneType") to str
@given(st.sampled_from(MyFlag))
>   @example(MyFlag.ONE | MyFlag.TWO)

tests/unit/test_hypothesis_bug.py:13: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.venv/lib/python3.10/site-packages/hypothesis/core.py:1056: in _raise_to_user
    raise the_error_hypothesis_found
.venv/lib/python3.10/site-packages/hypothesis/core.py:470: in execute_explicit_examples
    bits = ", ".join(nicerepr(x) for x in arguments) + ", ".join(
.venv/lib/python3.10/site-packages/hypothesis/core.py:471: in <genexpr>
    f"{k}={nicerepr(v)}" for k, v in example_kwargs.items()
.venv/lib/python3.10/site-packages/hypothesis/internal/reflection.py:452: in nicerepr
    return re.sub(r"(\[)~([A-Z][a-z]*\])", r"\g<1>\g<2>", pretty(v))
.venv/lib/python3.10/site-packages/hypothesis/vendor/pretty.py:101: in pretty
    printer.pretty(obj)
.venv/lib/python3.10/site-packages/hypothesis/vendor/pretty.py:199: in pretty
    return printer(obj, self, cycle)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

obj = <MyFlag.TWO|ONE: 3>
p = <hypothesis.vendor.pretty.RepresentationPrinter object at 0x7fc633f41090>
cycle = False

    def _repr_enum(obj, p, cycle):
>       p.text(type(obj).__name__ + "." + obj.name)
E       TypeError: can only concatenate str (not "NoneType") to str

.venv/lib/python3.10/site-packages/hypothesis/vendor/pretty.py:804: TypeError
```
Root cause: mixed-value flags don't have a name.

Python 3.10
hypothesis 6.82.0

### PR DESCRIPTION
**Pull request by Zac-HD**
Closes #3709.

And since I was pushing some commits, also fixes #3707.

### CODE REVIEW

File: hypothesis-python/src/hypothesis/vendor/pretty.py [hunk 1]
```diff
@@ -801,7 +801,16 @@ def _repr_dataframe(obj, p, cycle):  # pragma: no cover
...
 
 def _repr_enum(obj, p, cycle):
-    p.text(type(obj).__name__ + "." + obj.name)
+    if obj.name:  # Single enum item
+        return p.text(f"{type(obj).__name__}.{obj.name}")
+
+    try:  # combination of flags
+        match = re.fullmatch(rf"<{type(obj).__name__}.(\S+?): .+>", repr(obj))
+        names = sorted(match.group(1).split("|"), key=type(obj)._member_names_.index)
+        p.text(" | ".join(f"{type(obj).__name__}.{x}" for x in names))
```
Comment (jobh) [inline hypothesis-python/src/hypothesis/vendor/pretty.py:810]: Might it be simpler to decompose the enum up-front? I.e.,
```
decomposed = [e for e in list(type(obj)) if e & obj]
repr = " | ".join([f"{type(obj).__name__}.{d.name}" for d in decomposed])
```

**Review by jobh — APPROVED** [review]
Looks good!

File: hypothesis-python/src/hypothesis/vendor/pretty.py [hunk 2]
```diff
@@ -801,7 +802,11 @@ def _repr_dataframe(obj, p, cycle):  # pragma: no cover
 
 
 def _repr_enum(obj, p, cycle):
-    p.text(type(obj).__name__ + "." + obj.name)
+    tname = type(obj).__name__
+    if isinstance(obj, Flag):
+        p.text(" | ".join(f"{tname}.{x.name}" for x in type(obj) if x & obj == x))
```
Comment (jobh) [inline hypothesis-python/src/hypothesis/vendor/pretty.py:807]: @Zac-HD I just realized is is possible to construct empty flags e.g. `Options(0)`. These are unnamed, and not constructible be name. Maybe
```
p.text(" | ".join(...) or f"{tname}(obj.value)")
```
Too late for this PR, sorry.

Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/vendor/pretty.py:807]: Oooh, wow, that would have been _bizzare_ if anyone hit it in the wild!  Will fix.

**Pull request merged**